types.go

```diff
@@ -1,16 +1,114 @@
 package dmetering
 
-// Event represents a metering event that needs to be recorded
-type Event struct {
-	// Defines the service that emitted the event (firehose, substreams, evmx ...)
-	Service string `json:"service"`
-	// Defines the method within the service  (grpc_blocks, grpc_run, http_get_state, ....)
-	Method         string            `json:"method"`
-	Metadata       map[string]string `json:"metadata"`
-	RequestsCount  uint64            `json:"requests_count,omitempty"`
-	ResponsesCount uint64            `json:"responses_count,omitempty"`
-	IngressBytes   float64           `json:"ingress_bytes,omitempty"`
-	EgressBytes    float64           `json:"egress_bytes,omitempty"`
-	WrittenBytes   float64           `json:"written_bytes,omitempty"`
-	ReadBytes      float64           `json:"read_bytes,omitempty"`
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	pbmetering "github.com/streamingfast/dmetering/pb/sf/metering/v1"
+
+	"github.com/streamingfast/dgrpc"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+func init() {
+	Register("grpc", func(config string, logger *zap.Logger) (EventEmitter, error) {
+		u, err := url.Parse(config)
+		if err != nil {
+			return nil, err
+		}
+
+		vals := u.Query()
+		network := vals.Get("network")
+		if network == "" {
+			return nil, fmt.Errorf("missing networkId query param to metering config")
+		}
+
+		endpoint := u.Host
+		if endpoint == "" {
+			return nil, fmt.Errorf("endpoint not specified (as hostname)")
+		}
+
+		return newGRPCEmitter(network, endpoint, logger)
+	})
+
+	Register("logger", func(_ string, logger *zap.Logger) (EventEmitter, error) {
+		return newLoggerEmitter(logger), nil
+	})
+
+	Register("null", func(_ string, _ *zap.Logger) (EventEmitter, error) {
+		return newNullEmitter(), nil
+	})
+}
+
+type grpcEmitter struct {
+	client pbmetering.MeteringClient
+
+	network string
+	logger  *zap.Logger
+}
+
+func newGRPCEmitter(network, endpoint string, logger *zap.Logger) (EventEmitter, error) {
+	client, err := newMeteringClient(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create external gRPC client %w", err)
+	}
+
+	return &grpcEmitter{
+		client:  client,
+		network: network,
+		logger:  logger,
+	}, nil
+}
+
+func (g *grpcEmitter) Emit(ctx context.Context, ev Event) error {
+	pbevent := ev.ToProtoMeteringEvent()
+	pbevent.Network = g.network
+
+	if pbevent.Network == "" || pbevent.Endpoint == "" {
+		g.logger.Warn("events must contain network and endpoint. dropping event", zap.Object("event", ev))
+	}
+
+	_, err := g.client.Emit(ctx, pbevent)
+	if err != nil {
+		return fmt.Errorf("unable to emit event: %w", err)
+	}
+	return nil
+}
+
+func newMeteringClient(endpoint string) (pbmetering.MeteringClient, error) {
+	var dialOptions []grpc.DialOption
+	conn, err := dgrpc.NewExternalClient(endpoint, dialOptions...)
```

Comment:
`conn` close handling is missing, we should have a way to gracefully terminate the client connection.